src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -32342,9 +32284,312 @@ static void F1<T1>(T1 t1)
+            comp.VerifyDiagnostics(
+                // (4,30): error CS0701: 'T?' is not a valid constraint. A type used as a constraint must be an interface, a non-sealed class or a type parameter.
+                // class D<T> where T : struct, T? { }
+                Diagnostic(ErrorCode.ERR_BadBoundType, "T?").WithArguments("T?").WithLocation(4, 30),
+                // (2,9): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                // class B<T> where T : T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(2, 9),
+                // (3,9): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                // class C<T> where T : class, T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(3, 9),
+                // (5,9): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                // class E<T> where T : A, T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(5, 9));
+        }
+
+        [WorkItem(27289, "https://github.com/dotnet/roslyn/issues/27289")]
+        [Fact]
+        public void NullableTInConstraint_02()
+        {
+            var source =
+@"class A<T, U>
+    where U : T?
+{
+}
+class B<T, U>
+    where T : class
+    where U : T?
+{
+}
+class C<T, U>
+    where T : U?
+    where U : T?
+{
+}
+class D<T, U>
+    where T : class, U?
+    where U : class, T?
+{
+}
+class E<T, U>
+    where T : class, U
+    where U : T?
+{
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (10,9): error CS0454: Circular constraint dependency involving 'T' and 'U'
+                // class C<T, U>
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "U").WithLocation(10, 9),
+                // (15,9): error CS0454: Circular constraint dependency involving 'T' and 'U'
+                // class D<T, U>
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "U").WithLocation(15, 9),
+                // (20,9): error CS0454: Circular constraint dependency involving 'T' and 'U'
+                // class E<T, U>
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "U").WithLocation(20, 9));
+        }
+
+        [WorkItem(27289, "https://github.com/dotnet/roslyn/issues/27289")]
+        [Fact]
+        public void NullableTInConstraint_03()
+        {
+            var source =
+@"class A<T> where T : T, T? { }
+class B<U> where U : U?, U { }
+class C<V> where V : V?, V? { }";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (3,26): error CS0405: Duplicate constraint 'V' for type parameter 'V'
```

Comment:
>V [](start = 63, length = 1)

Should we say "Duplicate constraint 'V?'..." instead?

Reply:
Yes, top-level nullability should be included in the message. This requires allowing `TypeSymbolWithAnnotations` as a `Diagnostic` argument. I have that change in the next PR.

---
In reply to: [197943396](https://github.com/dotnet/roslyn/pull/27525#discussion_r197943396) [](ancestors = 197943396)